src/server.rs

```diff
@@ -0,0 +1,294 @@
+    }
+
+    /// Perform a ScuttleButt "handshake" with another UDP server.
+    pub fn gossip(&self, addr: impl Into<String>) -> Result<(), anyhow::Error> {
+        Ok(self.channel.send(ChannelMessage::Gossip(addr.into()))?)
+    }
+}
+
+/// UDP server for ScuttleButt communication.
+struct UdpServer {
+    channel: UnboundedReceiver<ChannelMessage>,
+    scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    socket: Arc<UdpSocket>,
+}
+
+impl UdpServer {
+    async fn new(
+        channel: UnboundedReceiver<ChannelMessage>,
+        scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    ) -> anyhow::Result<Self> {
+        let socket = {
+            let address = &scuttlebutt.lock().await.self_node_id;
+            Arc::new(UdpSocket::bind(address).await?)
+        };
+
+        Ok(Self {
+            scuttlebutt,
+            channel,
+            socket,
+        })
+    }
+
+    /// Listen for new ScuttleButt messages.
+    async fn run(&mut self) -> anyhow::Result<()> {
+        let mut buf = [0; BUF_SIZE];
+        loop {
+            tokio::select! {
+                result = self.socket.recv_from(&mut buf) => match result {
+                    Ok((len, addr)) => {
+                        let _ = self.process_package(addr, &buf[..len]).await;
+                    }
+                    Err(err) => return Err(err.into()),
+                },
+                message = self.channel.recv() => match message {
+                    Some(ChannelMessage::Gossip(addr)) => {
+                        let _ = self.gossip(addr).await;
+                    },
+                    Some(ChannelMessage::Shutdown) | None => break,
+                },
+            }
+        }
+        Ok(())
+    }
+
+    /// Process a single UDP packet.
+    async fn process_package(&self, addr: SocketAddr, data: &[u8]) -> anyhow::Result<()> {
+        // Handle gossip from other servers.
+        let message = bincode::deserialize(data)?;
+        let response = self.scuttlebutt.lock().await.process_message(message);
+
+        // Send reply if necessary.
+        if let Some(message) = response {
+            let message = bincode::serialize(&message)?;
+            self.socket.send_to(&message, addr).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Gossip to another UDP server.
+    async fn gossip(&self, addr: String) -> anyhow::Result<()> {
+        let syn = self.scuttlebutt.lock().await.create_syn_message();
+        let message = bincode::serialize(&syn)?;
+        let _ = self.socket.send_to(&message, addr).await?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+enum ChannelMessage {
+    Gossip(String),
+    Shutdown,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::future::Future;
+    use std::time::Duration;
+
+    use crate::ScuttleButtMessage;
+
+    async fn timeout<O>(future: impl Future<Output = O>) -> O {
+        tokio::time::timeout(Duration::from_millis(100), future)
+            .await
+            .unwrap()
+    }
```

Comment:
The timeout is never needed in cases where the test passes, but I think it's a good idea to make failure quick for development.

Reply:
That makes sense.